Adding GEP-709: Cross namespace references from Routes (#711)
kubernetes-sigs/gateway-api: site-src/geps/gep-709.md

# GEP-709: Cross Namespace References from Routes

* Issue URL: https://github.com/kubernetes-sigs/gateway-api/issues/709
* Status: Implementable

## TLDR

This GEP attempts to tackle both cross namespace forwarding and route inclusion.
These are closely related concepts that can be solved with a new ReferencePolicy
resource that enables app admins to describe where they trust references from.

## Motivation/User Journeys/Background

This GEP enables the following capabilities:

1. Retaining full control of Gateway and Routes in an infra namespace, while
   targeting apps in different namespaces.
1. Traffic splitting between Services in different namespaces.

> 1. Mesh overrides to target Services in different namespaces. 

Author: Relates to the other GEP in flight. Essentially the idea that as a consumer in namespace foo, I want to override my requests to service bar in a different namespace and send them to bar-canary. https://deploy-preview-715--kubernetes-sigs-gateway-api.netlify.app/geps/gep-713#policy-attachment-for-mesh

> 1. Mesh overrides to target Services in different namespaces. 

Reviewer: can we add the reference so people can see the link

> 1. Mesh overrides to target Services in different namespaces. 

Author: Good idea, added a link

1. Delegating paths or hostnames to Routes in other namespaces.

## API

Anytime we allow crossing a namespace boundary, we need to be very cautious. To
ensure that this feature is safe, we need to enforce a handshake mechanism that
requires resources in both namespaces to agree to this reference. To accomplish
that, a new ReferencePolicy resource should be introduced.

![Reference Policy](images/709-referencepolicy.png)

With this model, Routes would be able to directly reference Routes and Services

> in other namespaces. These references would only be considered valid if a

Author: Yep, agreed with all of that. Ended up moving route -> route to out of scope section.

ReferencePolicy in the target namespace explicitly allowed it. 

The following example shows how a HTTPRoute in namespace foo could reference
a Service in namespace bar. In this example a ReferencePolicy in the bar
namespace explicitly allows references to Services from HTTPRoutes in the foo
namespace.

```yaml
kind: HTTPRoute
metadata:
  name: foo
  namespace: foo
spec:
  rules:
  - matches:
    - path: /bar
    forwardTo:
      backend:
      - name: bar
        namespace: bar
---
kind: ReferencePolicy
metadata:
  name: bar
  namespace: bar
spec:

>   from:

Author: I think due to the additive nature of this API I'd rather have multiple resources than one large resource, but open to ideas. I'm concerned that a list of lists would become hard to manage.

>   from:

Reviewer 2: I think it's better to encourage people to have more, smaller objects rather than less, larger objects as well.

I note that the `from` and `two` stanzas are slices here though, but we don't have any discussion on how they should be combined. I'll add more here on the structs themselves.

  - group: networking.gateway.k8s.io
    kind: HTTPRoute
    namespace: foo
  to:
  - group: core
    kind: Service
```

### Benefits

* Conceptually similar to NetworkPolicy.
* A separate resource enables admins to restrict who can allow cross namespace

>   references.

Reviewer 3: No action required.
Do we need any specific commentary on our Security model that might be useful in future when we look back on this GEP? I don't think so. Food for thought for you and others.

Reviewer 2: If we do update anything, it's probably only to point out something like "the purpose of this is to allow the person who controls the referent object the requirement to accept that reference, explicitly, with the intent of making cross-namespace references more secure by default."

* Provides consistent way to control references to any resource from a Route.
* Can be extended in the future for additional use cases.
* A single ReferencePolicy resource can be used for a namespace in place of
  separate handshake config on each Service or Route resource.

### Implementation Details
* `ReferencePolicy.From` and `ReferencePolicy.To` would require at least 1 entry.
* `Group`, `Kind`, and `Namespace` would all be required fields in
  `ReferencePolicy.From`.
* `Group` and `Kind` would both be required fields in `ReferencePolicy.To`.
* There would be no way to further refine the `From` fields. By trusting a kind
  of resource in a namespace, you're effectively trusting anyone in that
  namespace that can administer that kind of resource. Specifying a name does
  not provide any real protection.
* Although we may want to consider adding a `Name` field to `ReferencePolicy.To`

>   in the future, it is not clear that it is necessary at this point.

Reviewer 3: Can we enumerate which resources are supported in `from.GVK` and those in `to.GVK` in this proposal?

Author: I think we need to be clear that this is only for Gateway API implementations, but due to the possibility of custom Route types or any other potential extension points I'm not sure we can have a complete list of allowed resources here.

Author: Added guidance for which resources are considered "core" here.

# GEP-709: Cross Namespace References from Routes

* Issue URL: https://github.com/kubernetes-sigs/gateway-api/issues/709
* Status: Implementable

> ## TLDR

Reviewer: Do we say very explicitly what fields and kinds are intended to be covered as part of "core" behavior in this proposal?

Also do we say that implementations may cover more Kinds and fields, but that is Custom behavior and there is no conformance there. (is this a good idea?)

Author: I've added this to the proposed spec. I think non-core kinds would fall under extended conformance since the expected behavior is clear, it's just a difference of which types we expect all implementations to support.

> This GEP attempts to tackle both cross namespace forwarding and route inclusion.

Reviewer: It doesn't seem like we are actually defining route inclusion in this GEP.

We should probably say that this is: prepare the API for route inclusion.

Reviewer: BTW if there are references, it would be good to include to define what route inclusion means.

Author: Unfortunately we only have feature request issues that can be solved with Route inclusion (#634, #695) and a couple design docs. For now I've just clarified this sentence and added a brief description of Route inclusion.

These are closely related concepts that can be solved with a new ReferencePolicy
resource that enables app admins to describe where they trust references from.

## Motivation/User Journeys/Background

This GEP enables the following capabilities:

> 1. Retaining full control of Gateway and Routes in an infra namespace, while

Reviewer: ```
1. Keep same namespace references simple.
```

Author: Added as part of intro.

   targeting apps in different namespaces.
1. Traffic splitting between Services in different namespaces.
1. Mesh overrides to target Services in different namespaces. 

## ReferencePolicy

Anytime we allow crossing a namespace boundary, we need to be very cautious. To
ensure that this feature is safe, we need to enforce a handshake mechanism that
requires resources in both namespaces to agree to this reference. To accomplish
that, a new ReferencePolicy resource should be introduced.

![Reference Policy](images/709-referencepolicy.png)

With this model, Routes would be able to directly reference Routes and Services
in other namespaces. These references would only be considered valid if a
ReferencePolicy in the target namespace explicitly allowed it. 

The following example shows how a HTTPRoute in namespace foo could reference
a Service in namespace bar. In this example a ReferencePolicy in the bar
namespace explicitly allows references to Services from HTTPRoutes in the foo
namespace.

```yaml
kind: HTTPRoute
metadata:
  name: foo
  namespace: foo
spec:
  rules:
  - matches:
    - path: /bar
    forwardTo:
      backend:
      - name: bar
        namespace: bar
---

> kind: ReferencePolicy

Reviewer: We might want to call this "InboundReferencePolicy" or something make the direction obvious

Author: That's a good idea. Any thoughts on this name change @youngnick @hbagdi @jpeach @howardjohn?

Reviewer 2: I'm okay with it, although it's a little long. I had discounted it before because I don't see us adding an `OutboundReferencePolicy` at any point, so I didn't think the disambiguation was necessary. But I don't object if others think it makes things clearer, I'm a little deep into this design to be able to have a good idea of how the name looks at first glance.

Reviewer: I think ReferencePolicy is find. (per discussion)

metadata:
  name: bar
  namespace: bar
spec:
  from:
  - group: networking.gateway.k8s.io
    kind: HTTPRoute
    namespace: foo
  to:
  - group: core
    kind: Service
```

### API
This proposed API is fairly straightforward, but comes with a couple notable
decisions:

1. Each ReferencePolicy only supports a single From and To section. Additional
   trust relationships can be modeled with additional ReferencePolicy resources.
1. Resource names are intentionally excluded from this policy for simplicity and

>    because they rarely provide any meaningful protection.

Reviewer: ... as the owner of the namespace can always rename resources/change the content of the resources to match a given policy.

Author: Good call, added this extra context.

1. A single Namespace is allowed per "From" struct. Although a selector would be
   more powerful it may encourage unnecessarily insecure configuration.

```go
// ReferencePolicy identifies cross namespace relationships that are trusted for
// Gateway API.
type ReferencePolicy struct {
    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`

    // Spec defines the desired state of ReferencePolicy.
    Spec ReferencePolicySpec `json:"spec,omitempty"`
}

// ReferencePolicySpec identifies a cross namespace relationship that is trusted
// for Gateway API.
type ReferencePolicySpec struct {
    // From describes the trusted namespaces and kinds that can reference the
    // resources described in "To".
    //
    // Support: Core
    //
    // +kubebuilder:validation:MinItems=1

>     From []ReferencePolicyFrom `json:"from"`

Reviewer 2: Both `from` and `to` are slices, but we don't have anything describing how multiple entries in the slice must be combined.

Seems like it would probably be something like "Each entry in `from` must be considered to be an additional place that references can be valid from, or to put this another way, entries must be combined using OR."

Similarly for the `to` stanza, I think that combining entries with OR is the only thing that makes sense. But regardless, we should have the behavior documented here.

Author: Good catch, this was not well defined. Added your suggestion verbatim.

    // To describes the resources that may be referenced by the resources
    // described in "From".
    //
    // Support: Core
    //
    // +kubebuilder:validation:MinItems=1
    To []ReferencePolicyTo `json:"to"`
}

// ReferencePolicyFrom describes trusted namespaces and kinds.
type ReferencePolicyFrom struct {
    // Group is the group of the referrent.
    //
    // Support: Core
    //
    // +kubebuilder:validation:MinLength=1
    // +kubebuilder:validation:MaxLength=253

>     Group string `json:"group"`

Reviewer 2: Because all three of these values are required, each `from` reference must fully specify a group, kind, and namespace. This is nice for implementers because it's very straightforward, but will this be flexible enough, or should we consider label selectors here? 

The downside of including label selectors is that you then need two sets of fields that are both optional, so the zero values of the fields and slices of this struct really matter. This way is much simpler.

Author: I've been tempted to introduce selectors here but so far have avoided it due to the increased complexity they add. In v1alpha1 I think we have had a few too many selectors and it became difficult to keep track of everything. Direct references can still be confusing, but I think they're generally easier to understand. One of my primary concerns with selectors is that their default is so open. Interpreted literally, an empty selector would accept references from all namespaces. 

So all of that to say that I agree with you. If we really _need_ selectors, we can always consider adding them in the future.

    // Kind is kind of the referrent.
    //
    // Support: Core
    //
    // +kubebuilder:validation:MinLength=1
    // +kubebuilder:validation:MaxLength=253
    Kind string `json:"kind"`

    // Namespace is the namespace of the referrent.
    //
    // Support: Core
    //
    // +kubebuilder:validation:MinLength=1
    // +kubebuilder:validation:MaxLength=253
    Namespace string `json:"namespace,omitempty"`
}

> // ReferencePolicyTo describes trusted kinds.

Reviewer: "ReferencePolicyTo describes what Kinds are allowed as targets of the references."

Reviewer: (Not a trusted kind, it's object that we own anyway)

Author: Much better, thanks!

> type ReferencePolicyTo struct {

Reviewer 4: Having a policy per-service will likely be a bottleneck in some scenarios. For example, if we have an egress proxy, and have an `external-namespace`. I probably just want to say `from: egress-gateway, to: any service in external-namespace`. I am not sure that is re-presentable here, without 1 ReferencePolicy per service?

Reviewer 4: Similarly would be `from: anywhere`

Author: Both `to` and `from` structs don't have names. This just represents trust from resources of type a in namespace foo to resources of type b within the same namespace as the ReferencePolicy. So in this case I think the `to: any service in external-namespace` is quite straightforward.

On the other hand `from: anywhere` is intentionally not allowed. I'm not entirely convinced it's a use case we want to encourage, but if it were, I think using a namespace selector instead of a string would be the appropriate way to do it. That does feel like a potential extension point, but I'm pretty hesitant to start with it unless there are compelling use cases.

Reviewer 2: Yeah, I think one of the key parts about this proposal is that it allows access *from* a Group/Kind and namespace *to* a Group/Kind and namespace (where the `to` namespace is always "the namespace the object is in".) I think that if we want to make it more flexible than "named namespace", it should be a label selector across namespaces only.

As it stands, we can do "Allow from a group of things in a list of namespaces" using these structs, since `from` is a list.

    // Group is the group of the referrent.
    //
    // Support: Core
    //
    // +kubebuilder:validation:MinLength=1
    // +kubebuilder:validation:MaxLength=253
    Group string `json:"group"`

    // Kind is kind of the referrent.
    //
    // Support: Core
    //
    // +kubebuilder:validation:MinLength=1
    // +kubebuilder:validation:MaxLength=253
    Kind string `json:"kind"`
}
```

### Benefits

* Conceptually similar to NetworkPolicy.

> * A separate resource enables admins to restrict who can allow cross namespace

Reviewer 4: Can you give an example of when this would actually be useful? I am having a hard time thinking of a scenario where I would want to disallow a namespace from exposing themselves to other namespaces

Author: I think this is conceptually similar to organizations who want to restrict the namespaces that can create Services of type LB and use quotas to achieve that. It's possible that you want users to be able to create Services without allowing them to expose them externally.

Reviewer 2: The use case for not allowing things to expose themselves to other namespaces is any cluster where a namespace admin for one namespace is not totally trusted. In that case, allowing inbound references is a critical step in privilege escalation (another reason to only allow this interaction in one direction).

For example, I've worked before on CI clusters, where each namespace ran a step in a CI pipeline. Now, CI pipelines are essentaily remote-code-execution-as-a-service, so those workloads are very untrusted. Would someone here be able to create a Service record? Possibly, and if so, it's almost certainly good to ensure that they cannot create cross-namespace references to their namespace.

This is a little contrived, I know, but I think the principle of "relatively untrusted namespace owner" is a pretty sound usecase for this overall.

Reviewer 4: Hmm, I am not sure I understand. Can you let me know where I am going wrong? This is my understanding:

Lets say we have a namespace `ingress`, which contains a Gateway. We have a namespace `untrusted`, which has a Service.

I think the statement here is we are trying to avoid letting `untrusted` expose their service through `ingress`' LoadBalancer. We are NOT trying to restrict `untrusted` from creating an HTTPRoute though (that is independent of this and driven by Gateway/Route selection agreement); rather, we are trying to restrict `ingress` from creating an HTTPRoute referencing `untrusted`.

So by restricting `untrusted` from creating a ReferencePolicy, we _think_ we have achieved this. In reality, the `ingress` namespace can just point to something in its own namespace that forwards to `untrusted` (there are many ways to do this, for simplicity lets just say they set up a dumb TCP proxy that forwards all requests to `untrusted`). As a result, I would say that the use of restrictive RBAC policy on ReferencePolicy does not achieve the goal of not allowing a Service to expose itself.

Reviewer 2: The point of the ReferencePolicy is to require the owner of the `untrusted` namespace to accept incoming references. If they can't create ReferencePolicy objects, then they can't accept incoming references, and so say, the admin can create a more-open HTTPRoute that may reference a lot of Services, but only allow some to complete the two-way handshake.

I agree that it's a bit convoluted, but the idea here is to put the control over incoming references into the hands of the person who wons the referent. The main advantage of doing this with a separate object is that it allows us to apply this pattern to things whose spec we can't change (like Service), while still keeping that control in the hands of the object's owner. I think that the RBAC advantage of having a separate object is a minor one at best compared to that, but it may prove useful to people who are extremely careful with their RBAC policies.

I probably overemphasised this usecase previously, sorry about that.

Reviewer 4: > The point of the ReferencePolicy is to require the owner of the untrusted namespace to accept incoming references. If they can't create ReferencePolicy objects, then they can't accept incoming references, and so say, the admin can create a more-open HTTPRoute that may reference a lot of Services, but only allow some to complete the two-way handshake.

I think I understand what you are discussing now, but I don't think its a good idea. The fact its possible to do this with the current design is no problem - obviously we cannot stop users from doing obscure things, but I wouldn't want to recommend this.

Instead of the admin "handshaking" every namespace, but then restricting their ability to handshake back, it seems like they should just not "handshake" in the first place?

Reviewer 4: Might be getting a bit side tracked here though :slightly_smiling_face:

Reviewer 2: Agreed. I think you make a fair point, that it's a pretty unlikely scenario that we already provide betters tools to deal with. But I think that having the ability to RBAC the ReferencePolicy object separately is an advantage, just not that big of one.

  references.
* Provides consistent way to control references to any resource from a Route.
* Can be extended in the future for additional use cases.
* A single ReferencePolicy resource can be used for a namespace in place of
  separate handshake config on each Service or Route resource.

#### Exceptions

> If traffic is originating from the same location as the configured Route (such

Reviewer: locations => "trusted isolation domain"

Author: Refactored this entire section, PTAL.

> as sidecars in some mesh implementations) implementations may choose to ignore

Reviewer: use MAY (capital)

Author: Refactored this entire section, PTAL.

> ReferencePolicy. This should only be done if:
> * Other mechanisms like NetworkPolicy can be used to effectively limit
>   cross-namespace references.
> * The implementation clearly documents that ReferencePolicy is not honored.

Reviewer 2: Small change to make this clearer:
```suggestion
ReferencePolicy. This may only be done if:
* Other mechanisms like NetworkPolicy are used to effectively limit
  cross-namespace references.
* The implementation must clearly document that ReferencePolicy is not honored.
```

I think that "may" is okay here because we are following it up with clarifying "must" clauses straight afterwards.

Reviewer: MAY, MUST, SHOULD should be capitalized so it's obvious we are using them in LOUD RFC TALK.

Reviewer 2: Yes, that's a great way to do it.

Author: Now that the CVE is live, I've reworked this entire section, including "LOUD RFC TALK" :), PTAL.

> This exception is very unlikely to apply to any ingress implementations of the

Reviewer 4: Generally agree, but this could be plausible in a single-tenant ingress. For example `ingress` namespace has all Gateways and Routes for the ingress. Some routes reference other namespace. There is no security risk here.

Author: How is there no security risk in that scenario? Would it be possible the implementation to know that it was always going to be deployed in that way?

Author: After discussing this a bit more, I think there are some instances where ingress implementations deployed in this way could potentially deserve an exception. Fundamentally the distinction is if the implementation would be subject to some other cross-namespace restrictions such as NetworkPolicy. I'll update this section to be a bit more clear.

Reviewer 2: I suggested a wording update here, I think that if we make that change, it covers this usecase already.

Author: Completely reworded this section, PTAL.

# GEP-709: Cross Namespace References from Routes

* Issue URL: https://github.com/kubernetes-sigs/gateway-api/issues/709
* Status: Implementable

## TLDR

This GEP attempts to enable cross namespace forwarding from Routes and provide a
way to simplify adding Route inclusion (Routes including other Routes) in the
future. These are closely related concepts that can be solved with a new
ReferencePolicy resource that enables app admins to describe where they trust
references from.

## Motivation/User Journeys/Background

This GEP keeps same namespace references simple while enabling the following
capabilities for cross namespace references:

1. Retaining full control of Gateway and Routes in an infra namespace, while
   targeting apps in different namespaces.
1. Traffic splitting between Services in different namespaces.
1. Mesh overrides to target Services in different namespaces. (For more info,
   see GEP [#713](https://github.com/kubernetes-sigs/gateway-api/issues/713))

## ReferencePolicy

> Anytime we allow crossing a namespace boundary, we need to be very cautious. To

Reviewer: ```
[very cautious](cite CVE)
```
